src/ekat/logging/ekat_logger.hpp

```diff
@@ -0,0 +1,158 @@
+  default to "debug."  Both are easy to change.
+
+  The LogMpiPolicy can restrict console output to only rank 0, and it will append
+  rank ids to log files and log names.
+
+  The LogFilePolicy determines whether or not the log will also write output to a file.
+  Files and console can have different logging levels.  For example, the
+  log level for the LogBasicFile policy could be "debug," while the level of the console
+  logger could be set to "info".  In this case, no debug messages will show up on the
+  console, but they will be in the log files.
+
+  In principle, separate modules, classes, etc. could have their own logger.
+  These loggers can share output files, too; see the spdlog wiki
+(https://github.com/gabime/spdlog/wiki/2.-Creating-loggers#creating-multiple-file-loggers-with-the-same-output-file) for an example.
+
+  The ekat::Logger class is a subclass of spdlog::logger class, to simplify the many
+  features of that library to the ones EKAT is most likely to use, while retaining all
+  of its functionality.
+*/
+template <typename LogFilePolicy = LogNoFile,    // see ekat_log_file.hpp
+          typename LogMpiPolicy = LogIgnoreRank> // see ekat_log_mpi.hpp
+class Logger : public spdlog::logger {
+  public:
+    Logger() = delete;
+
+    Logger(const std::string& log_name, const std::string& level_str="debug") :
+     spdlog::logger(LogMpiPolicy::name_append_rank(log_name))
+  {
+
+    const auto clevel = Log::level::from_str(level_str);
+    auto csink = std::make_shared<spdlog::sinks::stdout_color_sink_mt>();
+    csink->set_level(clevel);
+
+    std::string fname = LogMpiPolicy::name_append_rank(log_name) + "_logfile.txt";
+    auto fsink = LogFilePolicy::get_file_sink(fname);
+
+    this->sinks().push_back(csink);
+    this->sinks().push_back(fsink);
+
+    LogMpiPolicy::update_sinks(this->sinks());
+
+    this->set_level(clevel);
+  }
+
+  spdlog::sink_ptr get_console_sink() {
+    return (this->sinks().size() > 1 ? this->sinks()[0] : spdlog::sink_ptr(nullptr));
+  }
+
+  spdlog::sink_ptr get_file_sink() {
+    return (this->sinks().size() > 1 ? this->sinks()[1] : this->sinks()[0]);
+  }
+
+  template <typename FP = LogFilePolicy>
+  typename std::enable_if<FP::has_filename, std::string>::type
+  get_logfile_name() {
+    std::string result;
+    auto fsink = this->get_file_sink();
+    spdlog::sinks::basic_file_sink_mt* bptr =
+      dynamic_cast<spdlog::sinks::basic_file_sink_mt*>(fsink.get());
+    spdlog::sinks::rotating_file_sink_mt* rptr =
+      dynamic_cast<spdlog::sinks::rotating_file_sink_mt*>(fsink.get());
+    if (bptr) {
+      result = bptr->filename();
+    }
+    else if (rptr) {
+      result = rptr->filename();
+    }
+    return result;
+  }
+
+  template <typename FP = LogFilePolicy>
+  typename std::enable_if<!FP::has_filename, std::string>::type
+  get_logfile_name() {return "null";}
+};
+
+template <typename LogMpiPolicy>
+class Logger<LogSharedFile, LogMpiPolicy> : public spdlog::logger {
```

Comment:
This seems like a lot of code duplication, mostly just to eliminate the creation of the file sink. My suggestion would be to add a method to the main Logger class called something like `share(name,level)` or something like that, that returns a `Logger<LogSharedFile,LogMpiPolicy` pointer sharing the sink with `*this`. This method creates a logger internally, and sets the csink/fsink in it (can they also share the csink, if the clevel is the same?).

Note: this would require the ability to set the sinks, but I think that's relatively easy and short:
- turn `Logger()=delete;` into `Logger() = default`, and make it private;
- make `template<typename FP,typename MP> Logger<FP,MP>` a friend class (so that `*this` can access the copied log inside the `share(name,level)` method to set the sinks.

Alternatively to the previous 2 items, add a ctor to the main Logger class that takes a csink and a fsink, and directly sets them internally, no queestion asked, and use that in the `share(name,level)` method.

Reply:
> This seems like a lot of code duplication, mostly just to eliminate the creation of the file sink.

You're right. Thanks for this suggestion.  I was able to reproduce the behavior by just adding another constructor (rather than a whole specialized class).